modules/phase_field/src/ics/NestedBoundingBoxIC.C

```diff
@@ -0,0 +1,121 @@
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "NestedBoundingBoxIC.h"
+#include "MooseMesh.h"
+
+#include <iostream>
+
+registerMooseObject("PhaseFieldApp", NestedBoundingBoxIC);
+
+namespace sizeVector_def_nested
+{
+// Convenience function for sizing a vector to "n" given a vector with size 1 or "n"
+std::vector<Real>
+sizeVector(std::vector<Real> v, std::size_t size)
+{
+  if (v.size() == 1)
+    return std::vector<Real>(size, v[0]);
+  else
+    return v;
+}
+}
+
+template <>
+InputParameters
+validParams<NestedBoundingBoxIC>()
+{
+  InputParameters params = validParams<InitialCondition>();
+  params.addClassDescription("Specify variable values inside a list of nested boxes shaped "
+                             "axis-aligned regions defined by pairs of opposing corners");
+  params.addRequiredParam<std::vector<Point>>(
+      "smaller_coordinate_corners",
+      "For 1D, these are the left points; for 2D, these are the bottom left corners; for 3D, these "
+      "are the back bottom left corners. The format is (corner1_x corner1_y corner1_z corner2_x "
+      "corner2_y corner2_z ...), starting from the innermost box to the outermost box");
+  params.addRequiredParam<std::vector<Point>>(
+      "larger_coordinate_corners",
+      "For 1D, these are the right points; for 2D, these are the top right corners; for 3D, these "
+      "are the front top right corners. The format is (corner1_x corner1_y corner1_z corner2_x "
+      "corner2_y corner2_z ...), starting from the innermost box to the outermost box");
+  params.addRequiredParam<std::vector<Real>>("inside", "The value of the variable inside each box");
+
+  params.addParam<Real>("outside", 0.0, "The value of the variable outside the largest boxes");
+
+  params.addParam<Real>(
+      "int_width", 0.0, "The width of the diffuse interface. Set to 0 for sharp interface.");
+
+  params.addClassDescription("Allows setting the initial condition of the value of a field inside "
+                             "multiple nested bounding boxes.");
+  return params;
+}
+
+NestedBoundingBoxIC::NestedBoundingBoxIC(const InputParameters & parameters)
+  : InitialCondition(parameters),
+    _c1(getParam<std::vector<Point>>("smaller_coordinate_corners")),
+    _c2(getParam<std::vector<Point>>("larger_coordinate_corners")),
+    _nbox(_c1.size()),
+    _int_width(getParam<Real>("int_width")),
+    _dim(_fe_problem.mesh().dimension()),
+    _inside(sizeVector_def_nested::sizeVector(getParam<std::vector<Real>>("inside"), _nbox)),
+    _outside(getParam<Real>("outside"))
+{
+  // make sure inputs are the same length
+  if (_c2.size() != _nbox || _inside.size() != _nbox)
+    mooseError("vector inputs must all be the same size");
+}
+
+Real
+NestedBoundingBoxIC::value(const Point & p)
+{
+  Real value = _outside;
+
+  if (_int_width < 0.0)
+    mooseError("'int_width' should be non-negative");
+
+  if (_int_width == 0.0)
+  {
+    for (unsigned int b = 0; b < _nbox; ++b)
```

Comment:
Duplicated code? Perhaps you can factor this same algorithm out and reuse it for each bounding box-like IC. I would definitely support having a common bounding box IC base class (you could take over the existing BoundingBoxIC class and have each of these classes inherit from it.

Reply:
I will push the modified version for now, and if it still has lots of duplicated code, I will switch to using base class. (One problem is that the duplicated code mostly happens between the two new objects, rather than the current MultiBoundingBoxIC object.)

Reply:
Just trying to get you to think about your design. Let's get this in, and then you can continue to refine.